src/mdns/mdns_mdnssd.cpp

```diff
@@ -313,67 +349,124 @@ void PublisherMDnsSd::HandleServiceRegisterResult(DNSServiceRef         aService
+            RecordService(originalInstanceName.c_str(), aType, aServiceRef);
         }
         else
         {
-            otbrLog(OTBR_LOG_INFO, "MDNS remove service %s", aName);
-            DiscardService(aName, aType, aServiceRef);
+            DiscardService(originalInstanceName.c_str(), aType, aServiceRef);
         }
     }
     else
     {
-        otbrLog(OTBR_LOG_ERR, "Failed to register service %s: %s", aName, DNSErrorToString(aError));
-        DiscardService(aName, aType, aServiceRef);
+        otbrLog(OTBR_LOG_ERR, "[mdns] failed to register service %s: %s", originalInstanceName.c_str(),
+                DNSErrorToString(aError));
+        DiscardService(originalInstanceName.c_str(), aType, aServiceRef);
     }
+
+    if (mServiceHandler != nullptr)
+    {
+        // TODO: pass the renewed service instance name back to SRP server handler.
+        mServiceHandler(originalInstanceName.c_str(), aType, error, mServiceHandlerContext);
+    }
+
+exit:
+    return;
 }
 
 void PublisherMDnsSd::DiscardService(const char *aName, const char *aType, DNSServiceRef aServiceRef)
 {
-    for (Services::iterator it = mServices.begin(); it != mServices.end(); ++it)
+    ServiceIterator service = FindPublishedService(aName, aType);
+
+    if (service != mServices.end())
     {
-        if (!strncmp(it->mName, aName, sizeof(it->mName)) && !strncmp(it->mType, aType, sizeof(it->mType)))
-        {
-            assert(aServiceRef == nullptr || aServiceRef == it->mService);
-            mServices.erase(it);
-            DNSServiceRefDeallocate(aServiceRef);
-            aServiceRef = nullptr;
-            break;
-        }
-    }
+        assert(aServiceRef == nullptr || aServiceRef == service->mService);
 
-    assert(aServiceRef == nullptr);
+        otbrLog(OTBR_LOG_INFO, "[mdns] remove service ref %p", service->mService);
+
+        DNSServiceRefDeallocate(service->mService);
+        mServices.erase(service);
+    }
 }
 
 void PublisherMDnsSd::RecordService(const char *aName, const char *aType, DNSServiceRef aServiceRef)
 {
-    for (Services::iterator it = mServices.begin(); it != mServices.end(); ++it)
+    ServiceIterator service = FindPublishedService(aName, aType);
+
+    if (service == mServices.end())
     {
-        if (!strncmp(it->mName, aName, sizeof(it->mName)) && !strncmp(it->mType, aType, sizeof(it->mType)))
-        {
-            assert(aServiceRef == it->mService);
-            ExitNow();
-        }
+        Service newService;
+
+        otbrLog(OTBR_LOG_INFO, "[mdns] add service: %s.%s (ref: %p)", aName, aType, aServiceRef);
+
+        strcpy(newService.mName, aName);
+        strcpy(newService.mType, aType);
+        newService.mService = aServiceRef;
+        mServices.push_back(newService);
+    }
+    else
+    {
+        assert(service->mService == aServiceRef);
     }
+}
+
+otbrError PublisherMDnsSd::MakeTxtRecord(const TxtList &aTxtList, uint8_t *aTxtData, uint16_t &aTxtLength)
```

Comment:
This method seems to be general for a mDNS publishers, instead of being specific for PublisherMDnsSd.